test/integration/pause-resume-test.js

```diff
@@ -0,0 +1,246 @@
+  });
+};
+
+exports.tearDown = function(tearDownDone) {
+  const connection = this.connection;
+  if (!connection) {
+    tearDownDone();
+    return;
+  }
+  connection.on('end', function() {
+    tearDownDone();
+  });
+  connection.close();
+};
+
+// This test reads a large number of rows from the database.
+// At 1/4 of the rows, Request.pause() is called.
+// After a delay, Request.resume() is called.
+// The test verifies that:
+//  - No 'row' events are received during the pause.
+//  - The socket and the two transforms are stopped during the pause.
+//  - No large amounts of data are accumulated within the transforms during the pause.
+//  - No data is lost.
+exports.testLargeQuery = function(test) {
+  const debugMode = false;
+  const totalRows = 200000;                                // total number of rows to read
+  const delayTime = 500;                                   // pause delay time in ms
+  const connection = this.connection;
+  let request;
+  let rowsReceived = 0;
+  let failed = false;                                      // used to suppress further error messages
+  let paused = false;
+
+  connection.on('error', function(err) {
+    test.ifError(err);
+  });
+  openRequest();
+
+  function openRequest() {
+    const sql =                                            // recursive CTE to generate rows
+      'with cte1 as ' +
+        '(select 1 as i union all select i + 1 from cte1 where i < ' + totalRows + ') ' +
+      'select i from cte1 option (maxrecursion 0)';
+    request = new Request(sql, onRequestCompletion);
+    request.on('row', processRow);
+    connection.execSql(request);
+  }
+
+  function onRequestCompletion(err) {
+    test.ifError(err);
+    test.equal(rowsReceived, totalRows, 'Invalid row count.');
+    test.done();
+  }
+
+  function processRow(columns) {
+    if (paused) {
+      fail('Row received in paused state.');
+    }
+    rowsReceived++;
+    if (columns[0].value != rowsReceived) {
+      fail('Invalid row counter value, value=' + columns[0].value + ', expected=' + rowsReceived + '.');
+      return;
+    }
+    if (rowsReceived == Math.round(totalRows / 4)) {
+      pause();
+    }
+  }
+
+  function pause() {
+    if (debugMode) {
+      dumpStreamStates(connection);
+      console.log('Start pause.');
+    }
+    paused = true;
+    request.pause();
+    setTimeout(resume, delayTime);
+  }
+
+  function resume() {
+    if (debugMode) {
+      console.log('End pause.');
+      dumpStreamStates(connection);
+    }
+    verifyStreamStatesAfterPause();
+    paused = false;
+    request.resume();
+  }
+
+  function verifyStreamStatesAfterPause() {
+    const packetSize = connection.messageIo.packetSize();
+    const socketRs = connection.socket._readableState;
+    if (semver.gte(process.version, '0.12.18')) {
+      test.ok(!socketRs.flowing,
+        'Socket is not paused.');
+    }
+    test.ok(socketRs.length >= Math.min(socketRs.highWaterMark - 512, 0x4000),
```

Comment:
OK, but it will not make it much clearer, because these are more heuristic values, not exact science. :)